### PR TITLE
extmod/modbluetooth: Query-by-uuid, query-completion and read status.

### PR DESCRIPTION
**Pull request by jimmo**
This PR includes the functionality from #5933 #5906 #5976 (by @codefreax and @agners)
 - Query services/chars by UUID
 - Events for query service/char/desc completion
 - Event for read status

The main change from those PRs is that I've made the event no longer a bitfield, they're just sequential event IDs. There wasn't really any use for the `trigger` option (unlike the low-level stuff like pins or uarts, you don't need to be doing the filtering in C). This simplifies a few things. It's a breaking change, but we needed to do that anyway.

In order to make this work on btstack and nimble, the read status needed to be a separate event.

Query by UUID also reuses the existing gattc_discover_* methods with an optional final argument, rather than adding new ones. Also implemented for btstack.

Also added docs and updated the multitests and examples to use the new events. Also improved the reliability of the multitests a bit by updating the way they wait for events (no longer gets tripped up by multiple events arriving at the same time).

Multitests pass on the various combinations of unix-btstack, pybd-nimble and pybd-btstack.

### CODE REVIEW

**Comment by petrkr** [comment]
allow that "scan by uuid" skip scanning services if I already know uuid of char? So I can directly connect to Char?

Also change event ID would lead to users who does not read changelogs (most of) to stop working already done code. Which would lead to flash older firmware "which works". It's because those events are not exposed and users shall re-define them-self. If they would be exposed, it would not happen.

**Comment by jimmo** [comment]
> allow that "scan by uuid" skip scanning services if I already know uuid of char? So I can directly connect to Char?

Yes, you can set the handle range to 1-65535

```py
ble.gap_discover_characteristics(conn_handle, 1, 65535, char_uuid)
```

> 
> Also change event ID would lead to users who does not read changelogs (most of) to stop working already done code. Which would lead to flash older firmware "which works". It's because those events are not exposed and users shall re-define them-self. If they would be exposed, it would not happen.

I agree, but that's why we said it was an experimental API and subject to change.

"This module is still under development and its classes, functions, methods and constants are subject to change."

But this is a concern... I'm not sure adding the constants to the firmware now would help (i.e. hopefully we don't need to change it again). What if I added a bluetooth.py which users could optionally copy (or freeze) which provided these constants (but I guess the user needs to know to update it too?)

Is it really true that existing users discover that new firmware releases exist and start using them without seeing the changelog?

**Comment by petrkr** [comment]
> "This module is still under development and its classes, functions, methods and constants are subject to change."

That is right, so if there is time to change something, it's now, after "stable" release will be late to do such big and incompatible changes

> But this is a concern... I'm not sure adding the constants to the firmware now would help (i.e. hopefully we don't need to change it again). What if I added a bluetooth.py which users could optionally copy (or freeze) which provided these constants (but I guess the user needs to know to update it too?)
>
it would help as bluetooth.py is always frozen in firmware and you can not upload it separately as it contains binding to C classes


> Is it really true that existing users discover that new firmware releases exist and start using them without seeing the changelog?

since there is actually no change log, only bunch commits in git, then yes, I guess there are not many peoples who really read all 300+ commits between versions.
https://micropython.org/download/esp32/

**Comment by jimmo** [comment]
> it would help as bluetooth.py is always frozen in firmware and you can not upload it separately as it contains binding to C classes

The whole point was to keep these constants out of the default firmware. (So they don't take up space in ROM for people who don't use Bluetooth). When you include them inline in your code (using the underscore prefix and `const`), they don't even take up RAM (which wouldn't be the case if they were in bluetooth.py).

Perhaps this was a bad decision... in hindsight I kind of wish I'd named everything "svcs" "chars" "descs" rather than the full names too. I'll leave that up to @dpgeorge 

>since there is actually no change log, only bunch commits in git, then yes, I guess there are not many peoples who really read all 300+ commits between versions.

The change logs are at (e.g.) https://github.com/micropython/micropython/releases/tag/v1.12 which contains a summary of changes as well as the commit history.  This also gets cross-posted to the forum and (I think?) the newsletter.

> https://micropython.org/download/esp32/

Might be a good idea to link it here though too. I'll look into it, thanks.

File: extmod/nimble/modbluetooth_nimble.c [hunk 1]
```diff
@@ -763,6 +763,8 @@ STATIC int peripheral_discover_service_cb(uint16_t conn_handle, const struct ble
     if (error->status == 0) {
         mp_obj_bluetooth_uuid_t service_uuid = create_mp_uuid(&service->uuid);
         mp_bluetooth_gattc_on_primary_service_result(conn_handle, service->start_handle, service->end_handle, &service_uuid);
+    } else {
+        mp_bluetooth_gattc_on_discover_complete(MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE, conn_handle, error->status);
     }
```
Comment (friebel) [inline extmod/nimble/modbluetooth_nimble.c:768]: What's you opinion on the idea of reporting 0 instead of `BLE_HS_EDONE`? Thereby we could state in the API doc that "status is 0 if the discover operation completed successfully, or an implementation-specific error code otherwise".

File: extmod/modbluetooth.c [hunk 2]
```diff
@@ -886,6 +892,9 @@ STATIC mp_obj_t bluetooth_ble_invoke_irq(mp_obj_t none_in) {
         } else if (event == MP_BLUETOOTH_IRQ_GATTC_DESCRIPTOR_RESULT) {
             // conn_handle, handle, uuid
             ringbuf_extract(&o->ringbuf, data_tuple, 2, 0, NULL, 0, &o->irq_data_uuid, NULL);
+        } else if (event == MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE || event == MP_BLUETOOTH_IRQ_GATTC_CHARACTERISTICS_COMPLETE || event == MP_BLUETOOTH_IRQ_GATTC_DESCRIPTORS_COMPLETE) {
+            // conn_handle, status
+            ringbuf_extract(&o->ringbuf, data_tuple, 2, 0, NULL, 0, NULL, NULL);
         } else if (event == MP_BLUETOOTH_IRQ_GATTC_READ_RESULT || event == MP_BLUETOOTH_IRQ_GATTC_NOTIFY || event == MP_BLUETOOTH_IRQ_GATTC_INDICATE) {
```
Comment (friebel) [inline extmod/modbluetooth.c:898]: This belongs into the commit before?

File: extmod/nimble/modbluetooth_nimble.c [hunk 3]
```diff
@@ -74,19 +74,19 @@ STATIC int ble_hs_err_to_errno(int err) {
 }
 
 // Note: modbluetooth UUIDs store their data in LE.
-STATIC ble_uuid_t *create_nimble_uuid(const mp_obj_bluetooth_uuid_t *uuid) {
+STATIC ble_uuid_t *create_nimble_uuid(const mp_obj_bluetooth_uuid_t *uuid, ble_uuid_any_t *storage) {
     if (uuid->type == MP_BLUETOOTH_UUID_TYPE_16) {
```
Comment (friebel) [inline extmod/nimble/modbluetooth_nimble.c:78]: Instead of this, you could factor `init_nimble_uuid` out of `create_nimble_uuid`. What do you think?

File: extmod/nimble/nimble/npl_os.c [hunk 4]
```diff
@@ -127,6 +127,12 @@ void *nimble_realloc(void *ptr, size_t size) {
     return ptr2;
 }
 
+// No-op implementation (only used by NimBLE logging).
+int nimble_sprintf(char *str, const char *fmt, ...) {
+    str[0] = 0;
+    return 0;
+}
+
 /******************************************************************************/
```
Comment (friebel) [inline extmod/nimble/nimble/npl_os.c:136]: unrelated?

File: extmod/nimble/modbluetooth_nimble.c [hunk 5]
```diff
@@ -844,6 +844,7 @@ STATIC int ble_gatt_attr_read_cb(uint16_t conn_handle, const struct ble_gatt_err
     if (error->status == 0) {
         gattc_on_data_available(MP_BLUETOOTH_IRQ_GATTC_READ_RESULT, conn_handle, attr->handle, attr->om);
     }
+    mp_bluetooth_gattc_on_read_write_status(MP_BLUETOOTH_IRQ_GATTC_READ_STATUS, conn_handle, attr->handle, error->status);
```
Comment (friebel) [inline extmod/nimble/modbluetooth_nimble.c:847]: (just double-checking)
This posts a `READ_STATUS` for successful reads, too. Is this also the behavior of btstack?
Is the order of events (result first, then status) the same on btstack?

File: extmod/btstack/modbluetooth_btstack.c [hunk 6]
```diff
@@ -164,13 +164,13 @@ STATIC void btstack_packet_handler(uint8_t packet_type, uint8_t *packet, uint8_t
         DEBUG_EVENT_printf("  --> gatt query complete\n");
         uint16_t conn_handle = gatt_event_query_complete_get_handle(packet);
         uint16_t status = gatt_event_query_complete_get_att_status(packet);
-        if (irq == MP_BLUETOOTH_IRQ_GATTC_WRITE_STATUS) {
-            // TODO there is no value_handle to pass here
-            mp_bluetooth_gattc_on_write_status(conn_handle, 0, status);
-        } else if (irq == MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE || irq == MP_BLUETOOTH_IRQ_GATTC_CHARACTERISTICS_COMPLETE || irq == MP_BLUETOOTH_IRQ_GATTC_DESCRIPTORS_COMPLETE) {
```
Comment (friebel) [inline extmod/btstack/modbluetooth_btstack.c]: nit: this reformats a line introduced in a commit or two before

File: docs/library/ubluetooth.rst [hunk 7]
```diff
@@ -313,33 +331,35 @@ Central Role (GATT Client)
     Returns ``False`` if the connection handle wasn't connected, and ``True``
     otherwise.
 
-.. method:: BLE.gattc_discover_services(conn_handle)
+.. method:: BLE.gattc_discover_services(conn_handle, [uuid])
 
     Query a connected peripheral for its services.
 
+    Optionally specify
+
```
Comment (friebel) [inline docs/library/ubluetooth.rst:339]: Documentation of `uuid` incomplete/missing (also below)

**Review by friebel — COMMENTED** [review]
Nice, I think this is the right change!

File: extmod/modbluetooth.h [hunk 8]
```diff
@@ -88,48 +88,50 @@
...
+#define MP_BLUETOOTH_IRQ_CENTRAL_CONNECT                (0)
+#define MP_BLUETOOTH_IRQ_CENTRAL_DISCONNECT             (1)
+#define MP_BLUETOOTH_IRQ_GATTS_WRITE                    (2)
+#define MP_BLUETOOTH_IRQ_GATTS_READ_REQUEST             (3)
+#define MP_BLUETOOTH_IRQ_SCAN_RESULT                    (4)
+#define MP_BLUETOOTH_IRQ_SCAN_COMPLETE                  (5)
+#define MP_BLUETOOTH_IRQ_PERIPHERAL_CONNECT             (6)
+#define MP_BLUETOOTH_IRQ_PERIPHERAL_DISCONNECT          (7)
+#define MP_BLUETOOTH_IRQ_GATTC_SERVICE_RESULT           (8)
+#define MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE        (9)
```
[the remark this answers is not in the document]
Comment (jimmo) [reply extmod/modbluetooth.h:100]: I considered this, but seeing as this is the opportunity to re-order the numbers I figured I'd get them in now so they're grouped together.

File: extmod/nimble/modbluetooth_nimble.c (hunk 1 above)
Comment (jimmo) [reply extmod/nimble/modbluetooth_nimble.c:768]: Yes, I'm inclined to agree, good suggestion. Will fix.

File: extmod/modbluetooth.c (hunk 2 above)
Comment (jimmo) [reply extmod/modbluetooth.c:898]: Thanks, will fix.

File: extmod/nimble/modbluetooth_nimble.c (hunk 3 above)
Comment (jimmo) [reply extmod/nimble/modbluetooth_nimble.c:78]: Makes sense. I'll see what leads to the smallest code size and go with that.

File: extmod/nimble/nimble/npl_os.c (hunk 4 above)
Comment (jimmo) [reply extmod/nimble/nimble/npl_os.c:136]: Unfortunately `ble_gattc_disc_svc_by_uuid` calls `ble_gattc_log_disc_svc_uuid` which ultimately calls `ble_uuid_to_str` which calls `sprintf`.

File: extmod/nimble/modbluetooth_nimble.c (hunk 5 above)
Comment (jimmo) [reply extmod/nimble/modbluetooth_nimble.c:847]: Yeah, they come as separate events on btstack, and you only get the status on the "complete" event, so this was the easiest way to make it consistent between the two.

File: docs/library/ubluetooth.rst (hunk 7 above)
Comment (jimmo) [reply docs/library/ubluetooth.rst:339]: Thanks, will fix.

File: extmod/nimble/modbluetooth_nimble.c (hunk 1 above)
Comment (jimmo) [reply extmod/nimble/modbluetooth_nimble.c]: Done, and confirmed that this behaviour matches read/write status (and same for btstack).

File: docs/library/ubluetooth.rst [hunk 9]
```diff
@@ -313,33 +336,41 @@ Central Role (GATT Client)
...
-    will be raised.
+    will be raised, followed by ``_IRQ_GATTC_DESCRIPTORS_COMPLETE`` on completion.
 
 .. method:: BLE.gattc_read(conn_handle, value_handle)
 
     Issue a remote read to a connected peripheral for the specified
     characteristic or descriptor handle.
 
-    On success, the ``_IRQ_GATTC_READ_RESULT`` event will be raised.
+    When a value is available, the ``_IRQ_GATTC_READ_RESULT`` event will be raised. Additionally, the ``_IRQ_GATTC_READ_STATUS`` will be raised.
```
Comment (friebel) [inline docs/library/ubluetooth.rst:373]: nit: long line

**Review by friebel — APPROVED** [review]
LGTM!

**Comment by jimmo** [comment]
Updated to fix the docs, and rebased. Also added an additional commit to support gatt client reading characteristic values longer than 256 bytes which was previously being truncated. This requires `config(rxbuf=)` to be set of course.

**Comment by dpgeorge** [comment]
Thanks for this, it's a really good refinement of the bluetooth module.

Some points/notes/comments/questions:
- I think it's a good change to make events 8-bit integers instead of a bit field.  Just the saving of 1 byte in the ringbuf per event is worth it, because RAM is precious (actually more like 3 bytes saving because if we stuck with bit fields with more than 16 events it'd require 4 bytes in the ringbuf per event).  Yes this will be a small inconvenience for users to update, but they only need to change the cut-and-paste constants in their code, nothing else.
- Should events start at 1 instead of 0?  To reserve 0 as a special meaning (eg invalid/no event)?
- Copyright years should be updated to 2019-2020 (not changed from 2019 to 2020).
- Do we need to remove trigger from the irq() method? Maybe leave it for compatibility (with existing code, and with other classes like Pin.irq) and 0 means disable, non-zero means enable?
- `MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE`, perhaps rename it to `MP_BLUETOOTH_IRQ_GATTC_SERVICE_COMPLETE` (note removal of "S"), also for other new events?  This is just to make those result/complete events have the same `MP_BLUETOOTH_IRQ_GATTC_SERVICE_xxx` form.
- Is it worth having a single event for discovery complete, instead of three different events?  Or better to keep it and have more fine-grained information?
- The new READ_STATUS event: should it fire if status==0?  Is there a READ_RESULT if status is non-zero?  Does STATUS always follows RESULT?
- In the commit "extmod/modbluetooth: status=0 always on success" there are spurious (?) changes to the examples.

**Comment by petrkr** [comment]
>     * Is it worth having a single event for discovery complete, instead of three different events?  Or better to keep it and have more fine-grained information?


For us in blesync library would be also more better to have three separated events than making workarounds with one event. So if it is ok for you, go with three (I suppose you talking about _service _descriptor _characteristic events)

https://github.com/blesync/micropython-blesync/pull/7/commits/26fa2d295d6f18775b261d4e776dd75a5c882378

**Comment by dpgeorge** [comment]
> For us in blesync library would be also more better to have three separated events than making workarounds with one event. So if it is ok for you, go with three (I suppose you talking about _service _descriptor _characteristic events)

Yes I meant `MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE`, `MP_BLUETOOTH_IRQ_GATTC_CHARACTERISTICS_COMPLETE` and `MP_BLUETOOTH_IRQ_GATTC_DESCRIPTORS_COMPLETE`.  Since we can now have up to 256 events I think it's ok to have these as separate ones.

**Comment by dpgeorge** [comment]
> in hindsight I kind of wish I'd named everything "svcs" "chars" "descs" rather than the full names too.

Looking at the set of method names that currently exist they are (with possible renames):
```
active
config
irq
gap_advertise
gap_connect
gap_scan
gap_disconnect
gatts_register_services -> gatts_register_svcs
gatts_read
gatts_write
gatts_notify
gatts_set_buffer
gattc_discover_services -> gattc_discover_svcs
gattc_discover_characteristics -> gattc_discover_chars
gattc_discover_descriptors -> gattc_discover_descs
gattc_read
gattc_write
```
For consistency with other APIs (eg WLAN) we would want to keep the full words for `active`, `config`, `connect`, `disconnect`, `scan`, `read`, `write`.  After that there's not much left, and shortening just the svcs/chars/descs words doesn't really save that much.  Probably doing it from scratch I would shorten these but I don't think it's worth the hassle to change it now and break everything that uses BLE...

**Comment by dpgeorge** [comment]
In terms of backwards incompatibility, it seems the only thing users will need to change are the IRQ event codes.  The rest of the API is unchanged.  Is that correct?

If things were going to be changed more drastically I would suggest thinking about changing connect/disconnect events: combine the four of them (central/peripheral connect/disconnect) to just `_IRQ_CONNECT` and `_IRQ_DISCONNECT` with data `(conn_handle, addr_type, addr, master_slave)` and `(conn_handle, reason)` respectively.  This is because CONNECT isn't really central/peripheral based (a peripheral can initiate a connect to a central), and DISCONNECT can't tell between them without saving state about conn_handle that is disconnected (such state is actually saved by both nimble and btstack so that's not really a big issue here).

Also, from my tests it seems the DISCONNECT event can happen without a prior CONNECT event if trying to `gap_connect` to a BDADDR which is not advertising.  So logically I think it's better to have just CONNECT/DISCONNECT events.  But that would be a more significant breaking change.

**Comment by jimmo** [comment]
> * Should events start at 1 instead of 0?  To reserve 0 as a special meaning (eg invalid/no event)?

Done.

> * Copyright years should be updated to 2019-2020 (not changed from 2019 to 2020).

Done.

> * Do we need to remove trigger from the irq() method? Maybe leave it for compatibility (with existing code, and with other classes like Pin.irq) and 0 means disable, non-zero means enable?

I think if someone was using it, their code is already needing update for the IRQ numbers, and this will at least be a very early error raised. So if nothing else it's a good signal to go and check the docs.

> * `MP_BLUETOOTH_IRQ_GATTC_SERVICES_COMPLETE`, perhaps rename it to `MP_BLUETOOTH_IRQ_GATTC_SERVICE_COMPLETE` (note removal of "S"), also for other new events?  This is just to make those result/complete events have the same `MP_BLUETOOTH_IRQ_GATTC_SERVICE_xxx` form.

Done.

> * The new READ_STATUS event: should it fire if status==0?

I wondered this... I think it's better for consistency if the STATUS event always fires.

> Is there a READ_RESULT if status is non-zero? 

No.

I guess the READ_RESULT and READ_STATUS events could be combined into a single event, but this would mean that btstack would have to reassemble them.

> Does STATUS always follows RESULT?

Yes.

> * In the commit "extmod/modbluetooth: status=0 always on success" there are spurious (?) changes to the examples.

Fixed.

**Comment by schneider42** [comment]
`enqueue_irq()` still uses `ringbuf_peek16(&o->ringbuf) != MP_BLUETOOTH_IRQ_SCAN_RESULT)`. I guess it needs to use something like a `ringbuf_peek8()` function.

**Comment by jimmo** [comment]
> `enqueue_irq()` still uses `ringbuf_peek16(&o->ringbuf) != MP_BLUETOOTH_IRQ_SCAN_RESULT)`. I guess it needs to use something like a `ringbuf_peek8()` function.

Great catch, thank you. Fixed.

We need a test for this...

**Comment by dpgeorge** [comment]
The remaining thing I see here is naming consistency of the completion events: there are now the following events which indicate completion:
```
_IRQ_SCAN_COMPLETE (existing event)
_IRQ_GATTC_SERVICE_COMPLETE (new event)
_IRQ_GATTC_CHARACTERISTIC_COMPLETE (new event)
_IRQ_GATTC_DESCRIPTOR_COMPLETE (new event)
_IRQ_GATTC_READ_STATUS (new event)
_IRQ_GATTC_WRITE_STATUS (existing event)
```
The `READ_STATUS` and `WRITE_STATUS` should be renamed to match the others, like `READ_COMPLETE` and `WRITE_COMPLETE`.  And then I think it'd be good to take the chance to shorten the names to DONE instead of COMPLETE.  So the new names would be:
```
_IRQ_SCAN_DONE
_IRQ_GATTC_SERVICE_DONE
_IRQ_GATTC_CHARACTERISTIC_DONE
_IRQ_GATTC_DESCRIPTOR_DONE
_IRQ_GATTC_READ_DONE
_IRQ_GATTC_WRITE_DONE
```
This is only a very mild breaking change, because users can call these constants whatever they like.  These will just be the recommended names to use (eg in docs, examples).

**Comment by jimmo** [comment]
> The READ_STATUS and WRITE_STATUS should be renamed to match the others, like READ_COMPLETE and WRITE_COMPLETE. And then I think it'd be good to take the chance to shorten the names to DONE instead of COMPLETE.

Done.

File: docs/library/ubluetooth.rst [hunk 10]
```diff
@@ -135,21 +154,25 @@ Event Handling
...
+    _IRQ_GATTC_SERVICE_DONE         = const(9)
+    _IRQ_GATTC_CHARACTERISTIC_RESULT     = const(10)
+    _IRQ_GATTC_CHARACTERISTIC_DONE  = const(11)
+    _IRQ_GATTC_DESCRIPTOR_RESULT         = const(12)
+    _IRQ_GATTC_DESCRIPTOR_DONE      = const(13)
+    _IRQ_GATTC_READ_RESULT               = const(14)
+    _IRQ_GATTC_READ_DONE               = const(15)
+    _IRQ_GATTC_WRITE_DONE              = const(16)
+    _IRQ_GATTC_NOTIFY                    = const(17)
+    _IRQ_GATTC_INDICATE                  = const(18)
```
Comment (agners) [inline docs/library/ubluetooth.rst:175]: This set of const's is off by 1 now, tehy should start with 1 too.

**Comment by agners** [comment]
Other than that one comment this looks good! I also tested the patchset on a TinyPico and successfully used the UUID variants of `gattc_discover_services` and `gattc_discover_characteristics`. as well as making use of all the new events. Everything looks good so far. Thanks for your work!

**Comment by jimmo** [comment]
Thanks @agners -- I also took the opportunity to fix it to match the `black` formatting style.

**Comment by dpgeorge** [comment]
This is now merged in e6881f08292d03f089185718c131f543d095089b through 8b7ae4e099c18d839d1999c49dd8a1795bf9d1ae

Thanks a lot for the efforts!
